pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,217 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	apiserverserviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	clientset "k8s.io/client-go/kubernetes"
+	v1authentication "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+	"k8s.io/klog"
+)
+
+// defaultExpiration defines the duration of a TokenRequest in seconds.
+var defaultExpiration = int64(3600)
+
+type DynamicControllerClientBuilder struct {
+	// ClientConfig is a skeleton config to clone and use as the basis for each controller client
+	ClientConfig *restclient.Config
+
+	// CoreClient is used to provision service accounts if needed and watch for their associated tokens
+	// to construct a controller client
+	CoreClient v1core.CoreV1Interface
+
+	// AuthenticationClient is used to check API tokens to make sure they are valid before
+	// building a controller client from them
+	AuthenticationClient v1authentication.AuthenticationV1Interface
+
+	// Namespace is the namespace used to host the service accounts that will back the
+	// controllers.  It must be highly privileged namespace which normal users cannot inspect.
+	Namespace string
+
+	roundTripperFuncMap map[string]func(http.RoundTripper) http.RoundTripper
+
+	mutex sync.RWMutex
+
+	clock clock.Clock
+}
+
+func NewDynamicClientBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		roundTripperFuncMap:  map[string]func(http.RoundTripper) http.RoundTripper{},
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(saName string) (*restclient.Config, error) {
+	_, err := t.getOrCreateServiceAccount(saName)
+	if err != nil {
+		return nil, err
+	}
+
+	configCopy := constructClient(t.Namespace, saName, t.ClientConfig)
+
+	rt, ok := t.get(saName)
+	if ok {
+		configCopy.WrapTransport = rt
+	} else {
+		dur := time.Duration(defaultExpiration*20/100) * time.Second
+		if err != nil {
+			return nil, err
+		}
+
+		cachedTokenSource := transport.NewCachedTokenSource(&tokenSourceImpl{
+			clientConfig:       t.ClientConfig,
+			namespace:          t.Namespace,
+			serviceAccountName: saName,
+			coreClient:         t.CoreClient,
+			authClient:         t.AuthenticationClient,
+		}, dur)
+		configCopy.WrapTransport = transport.TokenSourceWrapTransport(cachedTokenSource)
+
+		t.set(saName, configCopy.WrapTransport)
+	}
+
+	return &configCopy, nil
+}
+
+func (t *DynamicControllerClientBuilder) ConfigOrDie(name string) *restclient.Config {
```

Comment:
Much of this duplication with SAControllerClientBuilder could be removed with a shared partial implementation and embedded struct, but we don't have to get into that here.